### PR TITLE
feat: add immersive dark mode on windows

### LINKED ISSUE #18119: Windows chrome and context menu should be aware of dark mode setting
### Preflight Checklist

* [x] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/master/CONTRIBUTING.md) for this project.
* [x] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/master/CODE_OF_CONDUCT.md) that this project adheres to.
* [x] I have searched the issue tracker for a feature request that matches the one I want to file, without success.

### Problem Description

Recent builds of Windows 10 support a "dark mode" that changes the chrome and context menus to fit better with the rest of the OS. https://www.digitaltrends.com/computing/how-to-enable-dark-mode-in-windows-10/ has a write up of how to enable this.

Here's a side-by-side of Explorer and the Quick Start app on Windows `10.0.17763.437`:

This is complementary to #15316 but not the whole solution:

 - it'd be great if the defaults "just worked" based on this setting, so app developers and users would get this benefit without needing to manage configuration themselves
 - some apps choose to roll their own themes, but they don't have control over the main menu or context menu without also rolling their own implementations of those and hiding the default chrome

To go into more detail, here's what a context menu looks like that respects the dark mode setting:

And here's the equivalent main menu and context menu in the Electron Quick Start when using Electron 5:

### Proposed Solution

I'm not familiar with how this works under the hood, but happy to go spelunking if this is something that interests the maintainers to gather more information.

### Alternatives Considered

As mentioned above, app developers can roll their own dark theme for menus and switch to use a chromeless `BrowserWindow`. That's fine for people who want to invest in that, but if there's lots of interest in the default chrome working better then maybe this is worth diving into?

### LINKED ISSUE #32913: [Feature Request]: Immersive Dark mode
### Preflight Checklist

- [X] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/main/CONTRIBUTING.md) for this project.
- [X] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/electron/electron/issues) for a feature request that matches the one I want to file, without success.

### Problem Description

Developer cannot change caption bar theme. It remains white even if windows theme is set to dark.

### Proposed Solution

When themesource in electron is set to dark, use dwm window attributes or check an example here: [Notepad++ Source code](https://github.com/notepad-plus-plus/notepad-plus-plus/tree/master/PowerEditor/src/DarkMode)

### Alternatives Considered

Check out in Notepad++, Sublime Text, Windows Explorer, MSYS Console...

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by mlaurencin**
#### Description of Change

Closes #32913, closes #18119.

A refactored implementation of dark mode for Windows, using official API for Windows 11 and undocumented flag on Win 10 >= 20H1. It allows Electron apps to set a dark title bar based on the system preference.

cc @dsanders11 @deermichel @ckerr 

#### Checklist

- [x] PR description included and stakeholders cc'd
- [ ] `npm test` passes
- [ ] relevant documentation is changed or added
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: Added immersive dark mode on Windows.

### CODE REVIEW

**Comment by dsanders11** [comment]
Been waiting for this. 😄

I've built it and played with it on Windows 10. However, it seems to suffer from the same problem outlined in #25373:

> There is still a wart: Changing themeSource dynamically when the window is already shown isn't perfect: the property is set, but the window still has to be poked before the change takes effect, e.g. changing focus or minimizing/restoring the window. The usual programmatic ways of forcing a window refresh don't seem to work here; perhaps this is my error but let's see about getting guidance for the right recipe to use.

Using the fiddle [from our docs on dark mode](https://www.electronjs.org/docs/latest/tutorial/dark-mode) I can confirm that's the behavior I see, the titlebar doesn't change until something forces a window refresh like resizing it, minimizing, blurring it, etc. It *does* however refresh correctly when it's set to system and you toggle the system setting.

Also, can release notes be updated to clarify it's for Windows?

**Comment by deermichel** [comment]
@dsanders11 yup that's a known issue - I spend quite some time trying to find a fix but wasn't able to come up with a working programmatic "force-redraw" (RedrawWindow API etc.). Feel free to give it a try ;). For some reason I only hit that in vanilla Electron, not in Skype - also it's only an issue when setting dark mode from within the app; afaik it works when changing the system dark mode in OS settings.

File: build/fuses/fuses.json5 [hunk 1]
```diff
@@ -7,5 +7,6 @@
   "node_options": "1",
   "node_cli_inspect": "1",
   "embedded_asar_integrity_validation": "0",
-  "only_load_app_from_asar": "0"
+  "only_load_app_from_asar": "0",
+  "windows_immersive_dark_mode": "1"
```
Comment (deermichel) [inline build/fuses/fuses.json5:11]: "0" before merging :)

**Review by deermichel — CHANGES_REQUESTED** [review]

**Comment by deermichel** [comment]
@mlaurencin after merging, we should also add the new fuse to [@electron/fuses](https://github.com/electron/fuses) :)

File: shell/browser/win/dark_mode.cc [hunk 2]
```diff
@@ -1,178 +1,51 @@
...
-  if (_AllowDarkModeForApp) {
-    _AllowDarkModeForApp(true);
-  } else if (_SetPreferredAppMode) {
-    _SetPreferredAppMode(AllowDark);
-  }
-  _RefreshImmersiveColorPolicyState();
-
-  // check to see if dark mode is currently enabled
-  g_darkModeEnabled = _ShouldAppsUseDarkMode() && !IsHighContrast();
+  return S_OK;
```
Comment (dsanders11) [inline shell/browser/win/dark_mode.cc:35]:
```suggestion
  // Toggle the nonclient area active state to force the titlebar to redraw
  SendMessage(hWnd, WM_NCACTIVATE, FALSE, 0);
  SendMessage(hWnd, WM_NCACTIVATE, TRUE, 0);

  return S_OK;
```

Comment (dsanders11) [reply shell/browser/win/dark_mode.cc:35]: @deermichel, in my testing this fixes the issue where the titlebar doesn't update. 🙂

It's a bit hackish, but it's functional. My gut tells me there's a more elegant way to accomplish this, and it also tells me there's a rabbit hole with Chromium since it handles a lot of the window messages. If there's a root cause anywhere, I'd wager it might be found in the Chromium codebase, in [`ui/views/win/hwnd_message_handler.cc`](https://source.chromium.org/chromium/chromium/src/+/main:ui/views/win/hwnd_message_handler.cc). Only way to know for sure would be some time testing the theme change out in a simple "Hello World" Windows app and see if there's a difference in behavior, and go from there. Unfortunately I don't have the time to run that down at the moment, but this works, and I didn't see any negative effects from it.

File: shell/browser/win/dark_mode.cc [hunk 3]
```diff
@@ -1,178 +1,51 @@
-// Copyright (c) 2020 Microsoft Inc. All rights reserved.
+// Copyright (c) 2022 Microsoft Inc. All rights reserved.
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE-CHROMIUM file.
 
 #include "shell/browser/win/dark_mode.h"
 
 #include <dwmapi.h>  // DwmSetWindowAttribute()
```
Comment (dsanders11) [inline shell/browser/win/dark_mode.cc:8]:
```suggestion
#include <uxtheme.h>  // SetWindowTheme()

```

File: shell/browser/win/dark_mode.cc [hunk 4]
```diff
@@ -1,178 +1,51 @@
...
 // This namespace contains code originally from
-// https://github.com/ysc3839/win32-darkmode/
-// governed by the MIT license and (c) Richard Yu
+// https://github.com/microsoft/terminal
+// governed by the MIT license and (c) Microsoft Corporation.
 namespace {
 
-// 1903 18362
-enum PreferredAppMode { Default, AllowDark, ForceDark, ForceLight, Max };
+// Use private Windows API to set window theme.
```
Comment (dsanders11) [inline shell/browser/win/dark_mode.cc:18]: What are we considering "private Windows API" at this point? It looks like `DWMWA_USE_IMMERSIVE_DARK_MODE` is documented (although only listed as being valid for Windows 11 builds). Seems the only thing not documented is `L"DarkMode_Explorer"` and `DWMWA_USE_IMMERSIVE_DARK_MODE_BEFORE_20H1`? Let's document it, in the hopes of this not needing to be behind a fuse?

File: shell/browser/win/dark_mode.cc [hunk 5]
```diff
@@ -1,178 +1,51 @@
...
+HRESULT TrySetWindowTheme(HWND hWnd, bool dark) {
+  HRESULT result =
+      SetWindowTheme(hWnd, dark ? DARK_MODE_STRING_NAME : L"", nullptr);
```
Comment (dsanders11) [inline shell/browser/win/dark_mode.cc:21]: Interestingly in my testing on an up-to-date Windows 10 machine, this wasn't necessary, and nothing changed if I removed it.

No suggestion or anything, just making a note.

**Comment by dsanders11** [comment]
Current implementation doesn't change the titlebar for DevTools windows, but I was able to get that working wtih a small change, see 63e4ef24ba275266643a07beabe895c4ae1ee60b for how I implemented it.

**Comment by deermichel** [comment]
@dsanders11 Thanks! I'll check your comments tmw or next week; unfortunately i cannot guarantee for anything but the initially proposed changes, because they were tested in the wild 🤞🏽

**Comment by dsanders11** [comment]
@deermichel, disregard the commit about doing all windows to get the DevTools windows. I tested it again and it didn't work like it did before. 🤔 I also realized that it wouldn't work for the case where you open DevTools after the theme change, since it wouldn't re-run that code when the DevTools opens. So, disregard that. 🙂 I'll just open a bug once this lands that DevTools won't have dark mode titlebar (would be nice, but not too important).

File: shell/browser/win/dark_mode.cc (hunk 4 above)
Comment (deermichel) [reply shell/browser/win/dark_mode.cc:18]: Interesting, back when I created the patch there was no Win11, so apparently they [made the value stable](https://docs.microsoft.com/en-us/windows/win32/api/dwmapi/ne-dwmapi-dwmwindowattribute). That makes the API public for Win11, on Win10 I assume we can expect no API breakages at this point - maybe this is already enough reason to drop the fuse (I want to check back with Releases WG to be double-sure). Not sure about `L"DarkMode_Explorer"` .. `Explorer` is listed as [example in the docs](https://docs.microsoft.com/en-us/windows/win32/api/uxtheme/nf-uxtheme-setwindowtheme) - so it's kinda document.

File: shell/browser/win/dark_mode.cc (hunk 5 above)
Comment (deermichel) [reply shell/browser/win/dark_mode.cc:21]: Hmm that's weird, no effect for me on Win11 either. I believe it had something to do with lists or menus - but it also might've been for an older Win10 version. If we remove it, we would get rid of the last "magic value" being used.

File: shell/browser/win/dark_mode.cc (hunk 2 above)
Comment (deermichel) [reply shell/browser/win/dark_mode.cc:35]: Can you try it on Win11 please? It seems to me that this has been fixed in Windows (probably that's why the property was not officially documented before).

**Comment by deermichel** [comment]
@dsanders11 @mlaurencin Let me suggest a path for going forward here:
- On Win11, there is a public API (that seems to work reliably) - so I wouldn't put the feature behind a flag, but enable it for all.
- On Win10, we can offer a fuse for users to enable it there with the risk of private/undocumented API usage
- or we don't offer the option on Win10 at all (not sure if Win11 adoption is that widespread tbh)

**Comment by mlaurencin** [comment]
> @dsanders11 @mlaurencin Let me suggest a path for going forward here:
> 
> * On Win11, there is a public API (that seems to work reliably) - so I wouldn't put the feature behind a flag, but enable it for all.
> * On Win10, we can offer a fuse for users to enable it there with the risk of private/undocumented API usage
> * or we don't offer the option on Win10 at all (not sure if Win11 adoption is that widespread tbh)

If I am following correctly, the API is currently public for Win11 and private for Win10, so it would be a matter of converting the fuse to only apply to Win10 if it were to be kept, right?

File: shell/browser/win/dark_mode.cc (hunk 5 above)
Comment (dsanders11) [reply shell/browser/win/dark_mode.cc:21]: Even with it, scrollbars, buttons, and multiple `selects` are light mode styling unless you set `:root { color-scheme: light dark; }` in the CSS for the page, at which point they style as expected. I've got a note to do a PR updating the dark mode tutorial to add that CSS, as the current tutorial renders with light mode buttons since it doesn't have that.

So I guess the only question would be if it's serving a purpose on older Win10, say 20H1 and older. I'll grab a VM to test that. If it seems to have no effect there either, then I vote we just drop it since it doesn't get us anything (that we know of). If it is needed on older Win10, I vote we either wrap it in an if guard, or just don't support dark mode for those older versions.

File: shell/browser/win/dark_mode.cc (hunk 2 above)
Comment (dsanders11) [reply shell/browser/win/dark_mode.cc:35]: Tried it on Win11, you're right, doesn't look like that toggling is necessary. We can wrap it with an if guard and only do it on Win10.

**Comment by MarshallOfSound** [comment]
> or we don't offer the option on Win10 at all (not sure if Win11 adoption is that widespread tbh)

I much prefer this option, it keeps the code cleaner and makes more sense for consumers. "This is a Windows 11 feature".

File: shell/browser/win/dark_mode.cc (hunk 5 above)
Comment (dsanders11) [reply shell/browser/win/dark_mode.cc:21]: I tested it on Windows 1903, which is the first build where dark mode functionality exists. Does appear that `L"DarkMode_Explorer"` makes no difference, it's not necessary to get the titlebar black like we want.

So I vote we drop it, put only Win10 behind the fuse, and only do the toggle titlebar code on Win10. 🙂

**Comment by dsanders11** [comment]
I much prefer leaving it possible to use on Win10 - which is going to be around for a long time to come. Not having the functionality at the moment is an eyesore, especially since the menu bar does respect dark mode, making the out of the box experience on Windows feel unpolished.

File: shell/browser/win/dark_mode.cc (hunk 2 above)
Comment (deermichel) [reply shell/browser/win/dark_mode.cc:35]: allright, i don't have a win10 instance around to verify - but i'm ok keeping that code for win10

File: shell/browser/win/dark_mode.cc (hunk 5 above)
Comment (deermichel) [reply shell/browser/win/dark_mode.cc:21]: ok, yeah if we don't see any difference, let's remove it and if ever an issue comes up, we can try adding it again

**Comment by deermichel** [comment]
I agree with @dsanders11. 
- Dark mode has been around for a long time with widespread usage on Win10 - users might not understand why this should be a Win11-only feature now. 
- Our Win11 userbase is around 10% (compared to all Windows versions) and a quick google reveals adoption doesn't look much better on a larger scale. Do we want to withhold a "great" experience from 80%+ of Windows users?

For Win11, enabling the feature per default is a no-brainer. For Win10, putting it behind a fuse seems like a reasonable way to make this feature available to the public without risking too much (it is super unlikely the API will change on Win10 anyway as it is in maintenance mode & would even break msft apps).

**Comment by hipstersmoothie** [comment]
How close are we to getting this landed? I really don't want to create my own custom build of electron

**Comment by dsanders11** [comment]
@deermichel, take a look at my branch, hopefully we can grab these changes and finish this PR off: https://github.com/dsanders11/electron/commits/immersive-dark-mode-add-2

I changed the fuse to mention Win10 specifically, added a runtime check for Win11, and separated out the implementations so it's cleaner and clear which one is using private APIs.

**Comment by deermichel** [comment]
@dsanders11 thank you very much, i'll take a look next week

**Review by ckerr — APPROVED** [review]
@erickzhao asked if this requires API WG review. IMO it doesn't since I don't see any direct API changing, but since you _could_ make an argument that the feature flag is API:

API LGTM

**Review by jkleinsc — COMMENTED** [review]
API LGTM

**Review by MarshallOfSound — CHANGES_REQUESTED** [review]
I want to discuss this fuse before merge.  On PTO currently, back on Monday PST

**Comment by deermichel** [comment]
@MarshallOfSound please drop your comments today, so we can move this forward :)

**Comment by MarshallOfSound** [comment]
Sorry @deermichel I thought I commented on this, probably got lost in airport wifi somewhere.

Basically here's my thoughts, we put things behind a fuse for basically one reason "It's something we want to enforce via code signature".  Basically every fuse is a security related feature.  This is a diversion from that.  If this can be done via a runtime check it should probably be a `base::Feature` instead of a fuse.

But at the heart of it, I don't like the direction of "this works by default on windows 11, and on windows 10 but only if you enable this one weird feature that you might find in our docs somewhere".  In my book this either goes to just Windows 11 or Windows 11 and 10 because the risk is acceptable.  My understanding was the issue was runtime linkages might break if this API becomes unstable in 10, if we're confident enough to _always_ link to these methods then the runtime toggle doesn't matter at all because regardless of the runtime toggle we've linked / loaded these symbols.

TLDR:
* Windows 11 only
* Remove the flag completely and enable for 11 + 10

Choose one 🤷 I'm big -1 on this thing landing in it's current fused / flagged state as it's **very** self-serving, no one is going to turn it on other than like three apps 🤷

**Comment by dsanders11** [comment]
@deermichel, @MarshallOfSound, [how about this](https://github.com/dsanders11/electron/commit/f3911f117d5a400b8b8bf1d7e588dfe9679d444b)? It drops the fuse and only enables dark mode on Win 10 >= 20H1, where the flag value is the same as the documented value on Win 11. That should cover [90% of Windows 10 users](https://reports.adduplex.com/#/r/2022-04).

> My understanding was the issue was runtime linkages might break if this API becomes unstable in 10, if we're confident enough to always link to these methods then the runtime toggle doesn't matter at all because regardless of the runtime toggle we've linked / loaded these symbols.

The current implementation doesn't have any extra linkage considerations AFAIK, it's making use of a stable API with an undocumented flag value, which became documented in Win 11. If we limit the Win 10 versions to the ones where the flag value is the same as the documented value (>= 20H1), my feeling is there's very little risk involved and that's something we can ship.

**Comment by deermichel** [comment]
@dsanders11 agreed. Let's remove the fuse and proceed with Win 10 & 11 as proposed. There is no active feature work done on Windows 10 anymore and dwm style attributes are certainly nothing security-related that might be changed in future, so I don't see any risk here. (Also I want to get this off the todo list...)

**Comment by dsanders11** [comment]
@deermichel, can you pull f3911f117d5a400b8b8bf1d7e588dfe9679d444b onto this branch if you're happy with it? That way it's ready for review again.

**Comment by deermichel** [comment]
@dsanders11 done. thanks a ton! can we get this merged now? cc @MarshallOfSound @ckerr @mlaurencin

**Review by MarshallOfSound — APPROVED** [review]

**Review by miniak — APPROVED** [review]

**Comment by damienallen** [comment]
Very excited to test this out in the next release!
